seb_openedx/middleware.py

```diff
@@ -1,27 +1,132 @@
 """ Middleware for seb_openedx """
 
 import sys
+import inspect
+from django.http import HttpResponseNotFound
 from django.utils.deprecation import MiddlewareMixin
 from django.conf import settings
-from opaque_keys.edx.keys import CourseKey
-from seb_openedx.edxapp_wrapper.edxmako_module import render_to_response
+from opaque_keys.edx.keys import CourseKey, UsageKey
+from web_fragments.fragment import Fragment
+from seb_openedx.edxapp_wrapper.edxmako_module import render_to_response, render_to_string
+from seb_openedx.edxapp_wrapper.get_courseware_module import get_courseware_module
 from seb_openedx.permissions import get_enabled_permission_classes
+from seb_openedx.seb_courseware_index import SebCoursewareIndex
+from seb_openedx.edxapp_wrapper.get_xmodule_modulestore_module import get_xmodule_modulestore_module
+from seb_openedx.utils import unquote_slashes
+
+
+SEB_WHITELIST_PATHS = getattr(settings, 'SEB_WHITELIST_PATHS', [])
+SEB_BLACKLIST_CHAPTERS = getattr(settings, 'SEB_BLACKLIST_CHAPTERS', [])
 
 
 class SecureExamBrowserMiddleware(MiddlewareMixin):
     """ Middleware for seb_openedx """
 
-    # pylint: disable=inconsistent-return-statements
     def process_view(self, request, view_func, view_args, view_kwargs):
         """ Start point of to d4etermine cms or lms """
+        courseware = get_courseware_module()
         course_key_string = view_kwargs.get('course_key_string') or view_kwargs.get('course_id')
         course_key = CourseKey.from_string(course_key_string) if course_key_string else None
+        access_denied = False
+
         if course_key:
+            # By default is all denied
+            access_denied = True
+
+            if self.is_whitelisted_view(request, course_key):
+                # First: Broad white-listing
+                access_denied = False
+
+            if self.is_blacklisted_chapter(request, course_key):
+                # Second: Granular white-listing
+                access_denied = True
+
             active_comps = get_enabled_permission_classes()
             for permission in active_comps:
                 if permission().check(request, course_key):
-                    return
-            return render_to_response('seb-403.html', status=403)
+                    access_denied = False
+
+        if access_denied:
+            is_courseware_view = bool(view_func.__name__ == courseware.views.index.CoursewareIndex.__name__)
+            if is_courseware_view:
+                return self.courseware_error_response(request, *view_args, **view_kwargs)
+            return self.generic_error_response(request, course_key)
+        return None
+
+    def is_whitelisted_view(self, request, course_key):
+        """ First broad filter: whitelisting of paths/tabs """
+        views_module = inspect.getmodule(request.resolver_match.func).__name__
+
+        if 'discussion' in SEB_WHITELIST_PATHS and views_module == 'discussion.views':
```

Comment:
This complete function has a very high complexity derived from the likely to grow ammount of if statements. I think we can refactor it a bit so that most of what it does is straight forward compare some variables to its existence on the WHITELIST_PATHS list

Reply:
> What do you think about making this check if `views_module in SEB_WHITELIST_PATHS` and per default use `discussion.views` in the list?

Sure

> This complete function has a very high complexity derived from the likely to grow ammount of if statements. I think we can refactor it a bit so that most of what it does is straight forward compare some variables to its existence on the WHITELIST_PATHS list

The mental complexity is low due early returns, meaning any of those paths from line 56 to 68 end the function and the rest of the functions doesn't have to be understood if it stops there: and when it doesn't stop there then those lines don't matter and the logic occurrences between line 69 through 81

Reply:
For the second comment, the current complexity is not skyhigh, but I do see this becoming a problem when there are new `if`s and more checks. Which is very likely to happen after new edge cases are revealed during a long test run

Reply:
A lot of the logic here cannot be converted into simple strings in whitelist_path, for example `if 'courseware' in SEB_WHITELIST_PATHS and self.is_xblock_request(request):`is only for xblock requests made on the courseware

Reply:
the easy extension is using url_name, for example if you want to block      
 `[url(r'^course_specific_login/{}/$'.format(settings.COURSE_ID_PATTERN),external_auth_views.course_specific_login, name='course-specific-login'), url(r'^course_specific_register/{}/$'.format(settings.COURSE_ID_PATTERN), external_auth_views.course_specific_register, name='course-specific-register'),]` all you have to do is add "course-specific-" to SEB_WHITELIST_PATHS

The reason the other "ifs" are needed is because a lot of the urls we need to blocks don't have a name (or needs complex logic)

Reply:
I think that is very good, that is why I think only 3 checks are enough.
1 for courseware
2 for the urls
3 for the mapper dict
That would be brilliant